util/topgen/secure_prng.py

```diff
@@ -0,0 +1,321 @@
+
+        bitsleft = n_bits
+        R = 0
+        while (bitsleft > 0):
+            random_byte = int(self.fetchbyte())
+            if (bitsleft > 8):
+                R = (R << 8) | random_byte
+                bitsleft = bitsleft - 8
+            else:
+                random_byte >>= 8 - bitsleft
+                R = (R << bitsleft) | random_byte
+                bitsleft = 0
+        assert R >= 0
+        return (R)
+
+    def randbelow(self, n):
+        """Generates a random integer smaller than n.
+
+        Uses getrandbits() to generate enough random bits. Repeats until the
+        resulting number is smaller than n. The probability of this is more
+        than 50% in each loop iteration.
+        The amount of consumed randomness is not fixed.
+        """
+        length_bits = _ceil(_log(n, 2))
+        R = self.getrandbits(length_bits)
+        while R >= n:
+            R = self.getrandbits(length_bits)
+        return (R)
+
+    def shuffle(self, x):
+        """Shuffles a list x.
+
+        Uses the Fisher-Yates algorithm.
+        Each possible permutation of x is generated with equal probability.
+        """
+
+        length_list = len(x)
+        for i in reversed(range(length_list)):
+            j = self.randbelow(i + 1)
+            x[i], x[j] = x[j], x[i]
+
+    def choice(self, x):
+        """Randomly chooses an element from a list x.
+
+        Uses randbelow() to select an integer smaller than len(x) and returns
+        the corresponding element.
+        """
+
+        i = self.randbelow(len(x))
+        return (x[i])
+
+    def padded_hex(self, c, length):
+        """Convert integer to a hexadecimal string with leading zeros."""
+        payload = hex(c)[2:]
+        padding = '0' * (length - len(payload))
+        return '0x' + padding + payload
+
+    def test_point_gen(self, entropy_input):
+        """Generate a test point in NIST format
+
+        Generates data for a unit test.
+        1. Reseed CTR_DRBG with entropy_input.
+        2. Generate two words of 512 bits
+        3. Pack entropy_input, the second generated word and CTR_DRBG internal
+        state at each step into a test_point.
+        """
+        entropy = self.padded_hex(entropy_input, 64)
+        self.reseed(entropy_input)
+        K0 = self.padded_hex(self.Key, 32)
+        V0 = self.padded_hex(self.V, 32)
+        # The first generated word is not in the test vectors
+        self.getrandbits(512)
+        K1 = self.padded_hex(self.Key, 32)
+        V1 = self.padded_hex(self.V, 32)
+        test_word = self.padded_hex(self.getrandbits(512), 128)
+        K2 = self.padded_hex(self.Key, 32)
+        V2 = self.padded_hex(self.V, 32)
+        test_point = {
+            'EntropyInput': entropy,
+            'Key_Inst': K0,
+            'V_Inst': V0,
+            'Key_First': K1,
+            'V_First': V1,
+            'ReturnedBits': test_word,
+            'Key_Second': K2,
+            'V_Second': V2}
+        return test_point
```

Comment:
I am wondering if this function should better be moved to `secure_prng_test.py` as all the other test code.

Reply:
Unfortunately, this function needs to read instance variables `_inst.Key` and `_inst.V`. I would rather that the test continues using protected instance `_inst` because this is how the module will be used in practice.